### PR TITLE
Discrete Cosine reparameterizer

### LINKED ISSUE #2221: [FR] Automatic reparameterization
Could we implement automatic reparameterization? Maybe at least for autoguides?

A good test example would be Neal's funnel, as described in the paper.

See Gorinova, Moore, and Hoffman (2019)
https://arxiv.org/pdf/1906.03028.pdf

## Tasks
- [x] #2222 implement `poutine.param` and basic framework
- [x] #2235 generalize `LocScaleTransform` and implement an effect `pyro.infer.reparam.decenter()`
- [x] #2235 implement `TransformReparameterizer` (suggested by @fehiepsi ) ~~and an effect `pyro.infer.reparam.neutra()`~~
- [x] #2235 support `Independent` distributions
- ~~implement a Levy-Ito reparameterization~~
- ~~implement a `difference` or `cumsum` reparameterization for time series~~
- [ ] #2252 implement a Fourier transform reparameterization for time series

### PR DESCRIPTION
**Pull request by fritzo**
Resolves #2221 
Blocking #2253 
Joint work with @fehiepsi 

This implements a Discrete Cosine Transform reparameterizer, intended for use in time series models. The motivating model class is time series models where it is natural to specify a model in terms of increments (e.g. a Gaussian or StudentT or Stable process), but where that parameterization is extremely poorly conditioned, leading to heavily non-diagonal posterior covariances. By reparameterizing in the frequency domain we can nearly (sometimes exactly) diagonalize the posterior covariance.

Empirically this seems to work well for the stochastic volatility model in #2253, where log-volatility is a latent gaussian process. Using an `AutoDiagonalNormal` guide I get cleaner posterior median and lower loss.

## Tasks
- [x] implement low-level ops `dct` and `idct`
    @fritzo initially added torch-dct as a dependency.
    @fehiepsi then replaced this with a custom implementation.
- [x] test on a time series model, e.g. stochastic volatility #2253

### CODE REVIEW

**Comment by fehiepsi** [comment]
@fritzo You can use the following version.

```
def _real_of_complex_mul(a, b):
    ar, ai = a.unbind(-1)
    br, bi = b.unbind(-1)
    return ar * br - ai * bi

def dtc(x):
    # Ref: http://fourier.eng.hmc.edu/e161/lectures/dct/node2.html
    # Step 1
    y = torch.cat([x[..., ::2], x[..., 1::2].flip(-1)], dim=-1)
    # Step 2
    Y = torch.rfft(y, 1, onesided=False)
    # Step 3
    coef_real = torch.cos(torch.linspace(0, 0.5 * math.pi, x.size(-1) + 1))
    coef = torch.stack([coef_real[:-1], -coef_real[1:].flip(-1)], dim=-1)
    # XXX: scipy.fftpack.dct has a factor of 2
    return _real_of_complex_mul(coef, Y)


def idtc(x):
    # Step 1
    coef_real = torch.cos(torch.linspace(0, 0.5 * math.pi, x.size(-1) + 1))
    coef = torch.stack([coef_real[:-1], coef_real[1:].flip(-1)], dim=-1)
    Y = x.unsqueeze(-1) * coef
    # Step 2
    y = torch.irfft(Y, 1, onesided=False)
    # Step 3
    return torch.stack([y, y.flip(-1)], axis=-1).reshape(x.shape[:-1] + (-1,))[..., :x.size(-1)]
```

**Comment by fritzo** [comment]
@fehiepsi does that guarantee orthogonality, so that the Jacobian is 1? Do you think we should add those instead of depending on torch-dct?

**Comment by fehiepsi** [comment]
Oh, good point. We need that for jacobian. I will edit it soon.

**Comment by fritzo** [comment]
@fehiepsi If you'd like to use those versions, can you push to this branch and add a test against `scipy.fftpack.dct(..., norm="ortho")`?

File: pyro/ops/tensor_utils.py [hunk 1]
```diff
@@ -110,3 +110,48 @@ def repeated_matmul(M, n):
...
+    # In other words, Y = complex_mul(e^ik, X - i[0, X[:0:-1]])
+    xi = torch.nn.functional.pad(-x[..., 1:], (0, 1)).flip(-1)
+    X = torch.stack([x, xi], dim=-1)
+    coef_real = torch.cos(torch.linspace(0, 0.5 * math.pi, N + 1))
+    coef = torch.stack([coef_real[:-1], coef_real[1:].flip(-1)], dim=-1)
+    Y = _complex_mul(coef, X)
+    # Step 2
+    y = torch.irfft(Y, 1, onesided=False)
+    # Step 3
+    # FIXME: something is wrong here
```
Comment (fritzo) [inline pyro/ops/tensor_utils.py]: Is something still wrong here?

**Comment by fritzo** [comment]
Thanks @fehiepsi ! I have added docs to `dct` and `idct` and removed the dependency on `torch-dct`. I think this is ready for review. Can you in particular check `DiscreteCosineReparam.__call__()`?

File: pyro/ops/tensor_utils.py (hunk 1 above)
Comment (fehiepsi) [reply pyro/ops/tensor_utils.py]: Ah, it is legacy. Btw, it seems that padding methods will not work here (because `dct(padding(x))` will have a length different from x)... this is different from `convolve` where we apply fft and ifft in the same method.

Comment (fritzo) [reply pyro/ops/tensor_utils.py]: ok, thanks for thinking about it.
